### PR TITLE
Initial commit for web api sample demonstrating obo flow

### PR DESCRIPTION
**Pull request by ramya25**

### CODE REVIEW

File: msal-web-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthFilter.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                httpResponse.setStatus(500);
+                request.setAttribute("error", exc.getMessage());
+                request.getRequestDispatcher("/error").forward(request, response);
+                return;
+            }
+        }
+        chain.doFilter(request, response);
+    }
+
+    private boolean isAccessTokenExpired(HttpServletRequest httpRequest) {
```
Comment (kalyankrishna1) [inline msal-web-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthFilter.java]: >isAccessTokenExpired [](start = 20, length = 20)

Comment the method

File: msal-web-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthFilter.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        IAuthenticationResult result = AuthHelper.getAuthSessionObject(httpRequest);
+        return result.expiresOnDate().before(new Date());
+    }
+
+    private void processAuthenticationCodeRedirect(HttpServletRequest httpRequest, String currentUri, String fullUrl)
+            throws Throwable {
+
+        Map<String, List<String>> params = new HashMap<>();
+        for (String key : httpRequest.getParameterMap().keySet()) {
+            params.put(key, Collections.singletonList(httpRequest.getParameterMap().get(key)[0]));
```
Comment (kalyankrishna1) [inline msal-web-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthFilter.java]: >.get(key)[0] [](start = 83, length = 12)

What is in ".get(key)[0]", provide some comments

File: README.md [hunk 3]
```diff
@@ -1,64 +1,316 @@
...
+
+#### Configure the **msal-obo-sample** to use your Azure AD tenant
+
+Open `application.properties` in the src/main/resources folder. Fill in with your tenant and app registration information noted in the above registration step.
+
+- Replace *Enter_the_Tenant_Info_Here* with  Directory (tenant) ID.
+- *Enter_the_Application_Id_here* with the Application (client) ID.
+- *Enter_the_Client_Secret_Here* with the *key value* noted.
+
+#### Register the client Web app (Webapp-Openidconnect)
```
Comment (TiagoBrenck) [inline README.md]: Replace the title Webapp-Openidconnect, with the name that you re suggesting in the step

File: README.md [hunk 4]
```diff
@@ -1,64 +1,316 @@
...
+   - When you save this page, the key value will be displayed. Copy and save the value in a safe location. You'll need this key later to configure the project. This key value will not be displayed again, nor retrievable by any other means, so record it as soon as it is visible in the Azure portal.
+
+1. Select **API Permissions** blade:
+   - Click the **Add a permission** button,
+   - Select the **My APIs** tab,
+   - In the list of APIs, select the API `Web Api app`,
+   - In the **Delegated permissions** section, ensure that the right permissions are checked: **access_as_user**,
+   - Select the **Add permissions** button.
+
+#### Configure the **msal-web-sample** to use your Azure AD tenant
```
Comment (TiagoBrenck) [inline README.md]: Missing step to add KnownClientApplication on the web api manifest. You can read about it here: https://github.com/Azure-Samples/ms-identity-aspnet-webapi-onbehalfof#configure-known-client-applications-for-service-todolistservice-obo-sample-v2

File: README.md [hunk 5]
```diff
@@ -1,64 +1,316 @@
...
+    - Search "Connector Port"
+    - Replace "8080" with your desired port number
+
+Example: `http://localhost:8080/msal4jsample`
+
+### You're done, run the code
+
+Click on "Login" to start the process of logging in. Once logged in, you'll see the account information for the user that is logged in and a Button "Call OBO API" , which will call the Microsoft Graph API with the OBO token and display the basic information of the signed-in user. You'll then have the option to "Sign out".
+
+## About the Code
```
Comment (TiagoBrenck) [inline README.md]: There should be a section about the `/.default` scope, which is the responsible for the OBO to work.

File: README.md (hunk 4 above)
Comment (ramya25) [reply README.md]: I think that step is not necessary for this Java sample. We are not adding anything in the manifest

Comment (TiagoBrenck) [reply README.md]: You could check with @kalyankrishna1 but I think this is mandatory for the first consent screen.

File: README.md (hunk 3 above)
Comment (ramya25) [reply README.md]: replaced with the one I used.

File: README.md (hunk 4 above)
Comment (ramya25) [reply README.md]: added KnownClientApplication step

File: README.md (hunk 5 above)
Comment (ramya25) [reply README.md]: Added

File: README.md (hunk 4 above)
Comment (TiagoBrenck) [reply README.md]: @ramya25 I couldn't find the step to edit it in the manifest

Comment (ramya25) [reply README.md]: Sorry. I think Readme file in the last push is not updated. now updated with the missing steps

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.microsoft.azure</groupId>
+    <artifactId>ms-identity-java-webapi</artifactId>
+    <version>1.0-SNAPSHOT</version>
+
+    <modules>
```
Comment (kalyankrishna1) [inline pom.xml]: >modules [](start = 5, length = 7)

Contributing.md

File: msal-obo-sample/pom.xml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        <version>2.1.7.RELEASE</version>
+        <relativePath/> <!-- lookup parent from repository -->
+    </parent>
+    
+    <groupId>com.microsoft.azure</groupId>
+    <artifactId>msal-obo-sample</artifactId>
+    <version>1.0-SNAPSHOT</version>
+
+    <properties>
+        <java.version>1.8</java.version>
```
Comment (sangonzal) [inline msal-obo-sample/pom.xml]: I think we might not want to hardcode Java version here, as customers might be using different versions of Java

File: msal-obo-sample/pom.xml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
+            <artifactId>msal4j</artifactId>
+            <version>0.5.0-preview</version>
```
Comment (sangonzal) [inline msal-obo-sample/pom.xml]: We can bump this up to 1.0.0.

Comment (sangonzal) [reply msal-obo-sample/pom.xml]: These two comments (Java version and MSAL version) also apply to the msal-web-sample pom.xml

File: msal-obo-sample/pom.xml (hunk 7 above)
Comment (ramya25) [reply msal-obo-sample/pom.xml]: Removed java version

File: README.md [hunk 9]
```diff
@@ -1,64 +1,375 @@
...
+1. The Java Web API authorizes the caller and then calls another downstream Web API ([The Microsoft Graph](https://graph.microsoft.com)) after obtaining another [access token](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens) using the on-behalf-of flow.
+
+### Scenario
+
+This sample shows how to build a Java web app that uses the [OpenId Connect](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols) protocol to sign in and sign out an user and to get an access token for another Web API including for the Microsoft Graph using the MSAL4J library. It also shows how to use On-Behalf-Of flow to acquire another access token from a Wen API to call another API.
+For more information about how the protocols work in this scenario and other scenarios, see [Authentication Scenarios for Azure AD](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-authentication-scenarios),[Microsoft identity platform and OAuth 2.0 On-Behalf-Of flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow).
+
```
Comment (navyasric) [inline README.md]: This paragraph seems to be a repetition of what is explained very well in the overview section. I suggest we remove this paragraph. It is redundant and adding more text to be read.

File: README.md [hunk 10]
```diff
@@ -1,64 +1,375 @@
...
+---
+# A Java Web API that calls another web API with the Microsoft identity platform using the On-Behalf-Of flow
+
+## About this sample
+
+### Overview
+
+This sample demonstrates a Java web application calling a downstream Web API, which in turn calls the [Microsoft Graph](https://graph.microsoft.com) using an [access token](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens) obtained using the [on-behalf-of](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) flow. All these are secured using the [Microsoft identity platform (formerly Azure Active Directory for developers)](https://docs.microsoft.com/en-us/azure/active-directory/develop/).
+
+1. The Java web application uses the [Microsoft Authentication Library for Java (MSAL4J)](https://github.com/AzureAD/microsoft-authentication-library-for-java) to obtain an Access token the Microsoft identity platform for the authenticated user.
```
Comment (navyasric) [inline README.md]: typo: to obtain an access token from the Microsoft identity platform...

File: README.md [hunk 11]
```diff
@@ -1,64 +1,375 @@
...
+2. The access token is then used as a bearer token to authorize the caller in the Java Web API and then subsequently for Microsoft Graph API.
```
Comment (navyasric) [inline README.md]: rephrase as: "to authorize the caller in the Java web API and then subsequently exchanged for another access token for the Microsoft Graph API"

File: README.md [hunk 12]
```diff
@@ -1,64 +1,375 @@
...
+
+The flow is as follows:
+
+1. Sign-in the user in the client(web) application.
+1. Acquire an access token token for the Java Web API and call it.
```
Comment (navyasric) [inline README.md]: typo: token repeated twice

File: README.md [hunk 13]
```diff
@@ -1,64 +1,375 @@
...
+For more information about how the protocols work in this scenario and other scenarios, see [Authentication Scenarios for Azure AD](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-authentication-scenarios),[Microsoft identity platform and OAuth 2.0 On-Behalf-Of flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow).
+
+## How to run this sample
+
+To run this sample, you'll need:
+
+- Working installation of Java and Maven
+- An Internet connection
+- An Azure Active Directory (Azure AD) tenant. For more information on how to get an Azure AD tenant, see [How to get an Azure AD tenant](https://azure.microsoft.com/en-us/documentation/articles/active-directory-howto-tenant/)
+- A user account in your Azure AD tenant.
```
Comment (navyasric) [inline README.md]: I think we can remove the Internet connection and user account prerequisites since those are obvious.

File: README.md [hunk 14]
```diff
@@ -1,64 +1,375 @@
...
+    Important things to notice:
+
+    - The **scope** [.default](https://docs.microsoft.com/en-us/azure/active-directory/developv2-permissions-and-consent#the-default-scope) is a built-in scope for every application that refers to the static list of permissions configured on the application registration. In our scenario here, it enables the user to grant consent for permissions for both the Web API and the downstream API (Microsoft Graph). For example, the permissions for the Web API and the downstream API (Microsoft Graph) are listed below:
+             - Web Api sample (access_as_user)
+             - Microsoft Graph (user.read)
+
+    - When you use the `.default` scope, the end user is prompted for a combined set of permissions that include scopes from both the **Web Api** and **Microsoft Graph**.
+
+2. **SecurityResourceServerConfig** class
+
```
Comment (navyasric) [inline README.md]: Can we mention one line here that Spring Security is being used for token validation.

File: AppCreationScripts/sample.json [hunk 15]
```diff
@@ -1,7 +1,7 @@
 {
   "Sample": {
     "RepositoryUrl": "ms-identity-java-webapi",
-    "Title": "Integrating Azure AD into a Java web api application",
+    "Title": "This sample demonstrates a Java web app application calling a Java Web API that is secured using Azure Active Directory using the On-Behalf-Of flow",
```
Comment (navyasric) [inline AppCreationScripts/sample.json]: This sample demonstrates a Java web app calling a Java Web API which calls another downstream API (Microsoft Graph) using the On-Behalf-Of flow.

File: README.md (hunk 13 above)
Comment (ramya25) [reply README.md]: but mentioned as those are mandatory things which should be there to run the sample. Also all other sample's README has them.

**Pull request merged**